File: wallendar-app/app/src/main/java/com/td/wallendar/group/ui/GroupActivity.java
package com.td.wallendar.group.ui;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.td.wallendar.AbstractActivity;
import com.td.wallendar.R;
import com.td.wallendar.addcharge.ui.AddChargeActivity;
import com.td.wallendar.addmembers.ui.AddMembersActivity;
import com.td.wallendar.di.DependenciesContainer;
import com.td.wallendar.di.DependenciesContainerLocator;
import com.td.wallendar.group.GroupHistoryAdapter;
import com.td.wallendar.groupbalance.ui.GroupBalanceActivity;
import com.td.wallendar.groupmembers.ui.GroupMembersActivity;
import com.td.wallendar.models.Charge;
import com.td.wallendar.models.Group;
import com.td.wallendar.models.GroupHistory;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.util.List;

public class GroupActivity extends AbstractActivity implements GroupView {

    private static final int REQUEST_ADD_CHARGE = 1;
    private static final int REQUEST_ADD_MEMBERS = 2;
    private static final int REFRESH = 3;

    private GroupPresenter groupPresenter;
    private GroupHistoryAdapter groupHistoryAdapter;
    private RecyclerView recycler;

    private ExtendedFloatingActionButton addChargeFAB;
    private TextView groupTitle;
    // Intended to be nullable
    private Long groupId;

    private boolean needsToRefresh = false;

    private final String GROUP_ID = "GROUP_ID";
    private final String NEW_CHARGE = "NEW_CHARGE";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_group);
        groupId = getIntent().getExtras().getLong(GROUP_ID);
        setUpView();
        createPresenter();
    }

    private void setUpView() {
        groupHistoryAdapter = new GroupHistoryAdapter();

        recycler = findViewById(R.id.group_activity_recycler);
        recycler.setHasFixedSize(true);
        recycler.setLayoutManager(new LinearLayoutManager(getApplicationContext(), LinearLayoutManager.VERTICAL, false));
        recycler.setAdapter(groupHistoryAdapter);

        setUpAddChargeButton(groupId);

        groupTitle = findViewById(R.id.group_title);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);

        // TODO
        findViewById(R.id.group_events).setOnClickListener(view -> Toast.makeText(getApplicationContext(),
                getString(R.string.feature_not_ready), Toast.LENGTH_SHORT).show());
        findViewById(R.id.group_balances).setOnClickListener(view -> {
            final Intent intent = new Intent(this, GroupBalanceActivity.class);
            intent.putExtra(GROUP_ID, groupId);
            startActivityForResult(intent, REFRESH);
        });
        findViewById(R.id.group_members).setOnClickListener(view -> {
            final Intent intent = new Intent(this, GroupMembersActivity.class);
            intent.putExtra(GROUP_ID, groupId);
            startActivity(intent);
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_ADD_CHARGE && resultCode == RESULT_OK) {
            Charge charge = (Charge) data.getExtras().getSerializable(NEW_CHARGE);
            groupHistoryAdapter.addToDataset(charge);
            needsToRefresh = true;
        } else if (requestCode == REQUEST_ADD_MEMBERS && resultCode == RESULT_OK) {
            Toast.makeText(getApplicationContext(), getString(R.string.members_added_successful), Toast.LENGTH_SHORT).show();
        } else if (requestCode == REFRESH && resultCode == RESULT_OK) {

        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.add_members_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.add_members).setVisible(true);
        return true;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.add_members) {
            final Intent addMembersIntent = new Intent(this, AddMembersActivity.class);
            addMembersIntent.putExtra(GROUP_ID, groupId);
            startActivityForResult(addMembersIntent, REQUEST_ADD_MEMBERS);
            return true;
        }
        return false;
    }

    private void createPresenter() {
        groupPresenter = (GroupPresenter) getLastNonConfigurationInstance();

        if (groupPresenter == null) {
            final DependenciesContainer dependenciesContainer = DependenciesContainerLocator.locateComponent(this);
            final GroupsRepository groupsRepository = dependenciesContainer.getGroupsRepository();
            final SchedulerProvider schedulerProvider = dependenciesContainer.getSchedulerProvider();
            groupPresenter = new GroupPresenter(this, groupsRepository, schedulerProvider);
        }
    }

    @Nullable
    @Override
    public Object onRetainCustomNonConfigurationInstance() {
        return groupPresenter;
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    @Override
    public void bindGroup(Group group) {
        groupTitle.setText(group.getTitle());
    }

    @Override
    public void bindGroupHistory(List<GroupHistory> historic) {
        groupHistoryAdapter.setDataset(historic);
    }

    @Override
    public void onGroupLoadError() {
        Toast.makeText(getApplicationContext(), "There was an error loading the group, try again later", Toast.LENGTH_LONG).show();
    }

    private void setUpAddChargeButton(Long groupId) {
        addChargeFAB = findViewById(R.id.add_charge_fab);
        addChargeFAB.setOnClickListener(view -> {
            final Intent intent = new Intent(this, AddChargeActivity.class);
            intent.putExtra(GROUP_ID, groupId);
            startActivityForResult(intent, REQUEST_ADD_CHARGE);
        });
        // Shrink floating button when scrolling, extend at the top. Just fancy fab
        recycler.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    addChargeFAB.extend();
                }
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0 || dy < 0 && addChargeFAB.isExtended()) {
                    addChargeFAB.shrink();
                }
            }
        });
    }

    @Override
    public void onStart() {
        super.onStart();
        groupPresenter.onViewAttached(groupId);
    }

    @Override
    public void onStop() {
        super.onStop();
        groupPresenter.onViewDetached();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        groupPresenter.onViewDestroyed();
    }

    @Override
    public void onBackPressed() {
        if (needsToRefresh) {
            final Intent resultIntent = new Intent();
            setResult(RESULT_OK, resultIntent);
        }
        finish();
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/group/ui/GroupActivity.kt
package com.td.wallendar.group.ui

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.td.wallendar.AbstractActivity
import com.td.wallendar.R
import com.td.wallendar.addcharge.ui.AddChargeActivity
import com.td.wallendar.addmembers.ui.AddMembersActivity
import com.td.wallendar.di.DependenciesContainerLocator
import com.td.wallendar.group.GroupHistoryAdapter
import com.td.wallendar.groupbalance.ui.GroupBalanceActivity
import com.td.wallendar.groupmembers.ui.GroupMembersActivity
import com.td.wallendar.models.Charge
import com.td.wallendar.models.Group
import com.td.wallendar.models.GroupHistory

class GroupActivity : AbstractActivity(), GroupView {
    private var groupPresenter: GroupPresenter? = null
    private var groupHistoryAdapter: GroupHistoryAdapter? = null
    private var recycler: RecyclerView? = null
    private var addChargeFAB: ExtendedFloatingActionButton? = null
    private var groupTitle: TextView? = null

    // Intended to be nullable
    private var groupId: Long? = null
    private var isEvent: Boolean = false
    private var needsToRefresh = false
    private val GROUP_ID: String = "GROUP_ID"
    private val IS_EVENT: String = "IS_EVENT"

    private val NEW_CHARGE: String = "NEW_CHARGE"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_group)
        groupId = intent.extras?.getLong(GROUP_ID)
        isEvent = intent.extras?.getBoolean(IS_EVENT)!!
        setUpView()
        createPresenter()
    }

    private fun setUpView() {
        groupHistoryAdapter = GroupHistoryAdapter()
        recycler = findViewById(R.id.group_activity_recycler)
        recycler?.setHasFixedSize(true)
        recycler?.layoutManager = LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
        recycler?.adapter = groupHistoryAdapter
        setUpAddChargeButton(groupId)
        groupTitle = findViewById(R.id.group_title)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)

        // TODO
        findViewById<View?>(R.id.group_events).setOnClickListener {
            Toast.makeText(applicationContext,
                    getString(R.string.feature_not_ready), Toast.LENGTH_SHORT).show()
        }
        findViewById<View?>(R.id.group_events).visibility = if (isEvent) GONE else VISIBLE
        findViewById<LinearLayout?>(R.id.group_linear_buttons).weightSum = if (isEvent) 2F else 3F


        findViewById<View?>(R.id.group_balances).setOnClickListener {
            val intent = Intent(this, GroupBalanceActivity::class.java)
            intent.putExtra(GROUP_ID, groupId)
            intent.putExtra(IS_EVENT, isEvent)
            startActivityForResult(intent, REFRESH)
        }
        findViewById<View?>(R.id.group_members).setOnClickListener {
            val intent = Intent(this, GroupMembersActivity::class.java)
            intent.putExtra(GROUP_ID, groupId)
            intent.putExtra(IS_EVENT, isEvent)
            startActivity(intent)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_ADD_CHARGE && resultCode == RESULT_OK) {
            val charge = data?.extras?.getSerializable(NEW_CHARGE) as Charge?
            groupHistoryAdapter?.addToDataset(charge!!)
            needsToRefresh = true
        } else if (requestCode == REQUEST_ADD_MEMBERS && resultCode == RESULT_OK) {
            Toast.makeText(applicationContext, getString(R.string.members_added_successful), Toast.LENGTH_SHORT).show()
        } else if (requestCode == REFRESH && resultCode == RESULT_OK) {
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.add_members_menu, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        super.onPrepareOptionsMenu(menu)
        menu?.findItem(R.id.add_members)?.isVisible = true
        return true
    }

    @SuppressLint("NonConstantResourceId")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.add_members) {
            val addMembersIntent = Intent(this, AddMembersActivity::class.java)
            addMembersIntent.putExtra(GROUP_ID, groupId)
            startActivityForResult(addMembersIntent, REQUEST_ADD_MEMBERS)
            return true
        }
        return false
    }

    private fun createPresenter() {
        groupPresenter = lastNonConfigurationInstance as GroupPresenter?
        if (groupPresenter == null) {
            val dependenciesContainer = DependenciesContainerLocator.locateComponent(this)
            val groupsRepository = dependenciesContainer.getGroupsRepository()
            val schedulerProvider = dependenciesContainer.getSchedulerProvider()
            groupPresenter = GroupPresenter(this, groupsRepository, schedulerProvider)
        }
    }

    override fun onRetainCustomNonConfigurationInstance(): Any? {
        return groupPresenter
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    override fun bindGroup(group: Group) {
        groupTitle?.setText(group.title)
    }

    override fun bindGroupHistory(historic: MutableList<GroupHistory>) {
        groupHistoryAdapter?.setDataset(historic)
    }

    override fun onGroupLoadError() {
        Toast.makeText(applicationContext, "There was an error loading the group, try again later", Toast.LENGTH_LONG).show()
    }

    private fun setUpAddChargeButton(groupId: Long?) {
        addChargeFAB = findViewById(R.id.add_charge_fab)
        addChargeFAB?.setOnClickListener {
            val intent = Intent(this, AddChargeActivity::class.java)
            intent.putExtra(GROUP_ID, groupId)
            intent.putExtra(IS_EVENT, isEvent)
            startActivityForResult(intent, REQUEST_ADD_CHARGE)
        }
        // Shrink floating button when scrolling, extend at the top. Just fancy fab
        recycler?.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    addChargeFAB?.extend()
                }
                super.onScrollStateChanged(recyclerView, newState)
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0 || dy < 0 && addChargeFAB?.isExtended == true) {
                    addChargeFAB?.shrink()
                }
            }
        })
    }

    public override fun onStart() {
        super.onStart()
        groupPresenter?.onViewAttached(groupId!!)
    }

    public override fun onStop() {
        super.onStop()
        groupPresenter?.onViewDetached()
    }

    public override fun onDestroy() {
        super.onDestroy()
        groupPresenter?.onViewDestroyed()
    }

    override fun onBackPressed() {
        if (needsToRefresh) {
            val resultIntent = Intent()
            setResult(RESULT_OK, resultIntent)
        }
        finish()
    }

    companion object {
        private const val REQUEST_ADD_CHARGE = 1
        private const val REQUEST_ADD_MEMBERS = 2
        private const val REFRESH = 3
    }
}